### ISSUE TITLE
OpenAPI API definition integration

### ISSUE THREAD

**Issue by powelli** [serverless/serverless#8442]
### Use case description
In order to support interface-first design patterns, I would like to create an OpenAPI (previously Swagger) definition for the API I wish to deploy, then have Serverless:
1. Import this definition from an external file or inline
1. Enrich the API definition with integration config to connect API endpoints to Serverless functions (e.g. `x-amazon-apigateway-integration`)
1. Deploy the AWS API Gateway using the compiled definition

Potentially, this could be extended to allow the enriched API definition to be exported (e.g. to S3) for use in documentation.

This request builds upon an incremental feature discussed but deferred  under:
#7052

Other references/requests for this support can be found in the forum here:
https://forum.serverless.com/t/swagger-and-serverless-in-parallel-bad-idea/1095
https://forum.serverless.com/t/reference-to-a-external-swagger-api-definition/5633
https://forum.serverless.com/t/define-api-with-external-openapi-spec/11167

### Proposed solution
Taking the previously proposed configuration from #7052 - an implementation in OpenAPI & Serverless config could look like:

#### Original OpenAPI definition
```yaml
provider:
  httpApi:
    # Endpoints configuration provided via Open API object
    openApi: # can be imported as e.g. ${file(open-api.json)}
      openapi: "3.0.0"
      info:
        title: "HTTP API example"
        version: "1.0.0"
      paths:
        /users:
          get:
            operationId: usersGet
          post:
            operationId: usersPost
        /other:
          get:
            operationId: other
```

#### Serverless config
```
functions:
  # Individual HTTP route example
  users:
    handler: users.handler
    events:
      - httpApi:
          operationId: usersGet
     - httpApi:
          operationId: usersPost
  # Catch all other specified routes example
  api:
    handler: api.handler
    events:
      - httpApi:
          operationId: '*'
```

#### Generated OpenAPI definition
The above example files would generate a complete OpenAPI definition similar to:
```yaml
openapi: "3.0.0"
info:
  title: "HTTP API example"
  version: "1.0.0"
paths:
  /users:
    get:
      operationId: usersGet
      x-amazon-apigateway-integration:
        type: aws
        uri: arn:aws:apigateway:eu-west-2:lambda:path/2015-03-31/functions/arn:aws:lambda:eu-west-2:123456789:function:users/invocations
        credentials: arn:aws:iam::*:user/*
        httpMethod: GET
    post:
      operationId: usersPost
      x-amazon-apigateway-integration:
        type: aws
        uri: arn:aws:apigateway:eu-west-2:lambda:path/2015-03-31/functions/arn:aws:lambda:eu-west-2:123456789:function:users/invocations
        credentials: arn:aws:iam::*:user/*
        httpMethod: POST
  /other:
    get:
      operationId: other
      x-amazon-apigateway-integration:
        type: aws
        uri: arn:aws:apigateway:eu-west-2:lambda:path/2015-03-31/functions/arn:aws:lambda:eu-west-2:123456789:function:api/invocations
        credentials: arn:aws:iam::*:user/*
        httpMethod: GET
```

**Comment by medikoo** [comment]
@powelli great thanks for this proposal. We're definitely open to have that in. One question for start:

> The above example files would generate a complete OpenAPI definition similar to:

Can you elaborate on that. It'll generate it and expose internally? Save to some file? What would be the intended usage of expanded  OpenAPI definition (in simpler form maintained in local file) ?

**Comment by nponeccop** [comment]
> Can you elaborate on that

Although I'm not @medikoo, I can hopefully help here. The intended usage of the expanded definition is to put it into the `Body` property of  the `AWS::ApiGatewayV2::Api` resource. Now AFAIK this property is always undefined in CloudForge templates generated by Serverless.

Doing that lets you take advantage of various AWS OpenAPI extensions, including but not limited to the aforementioned `x-amazon-apigateway-integration`, as described in [Working with API Gateway extensions to OpenAPI][1]. 

The idea of the extensions is to eliminate duplication between the OpenAPI file and the CloudForge template: both these files have a list of API endpoints. Moreover, there are some OpenAPI features such as request schemas, which don't have a CloudForge equivalent.

So basically some of the AWS ApiGateway v2 aka `httpApi` features are inaccessible in Serverless now. And those are important features such as request validations performed by the AWS Lambda runtime, without clobbering your own code. Well, there are seem to be some ways to access them, e.g. by using the `Resources` to specify raw CloudForge properties directly, but it's possible to do better.

[1]: https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-swagger-extensions.html

**Comment by pgrzesik** [comment]
Hello @nponeccop - thanks for additional clarification on that. I'm currently diving a bit more to better understand how we could support OpenAPI for HTTP APIs in the Framework and having such a PoC with currently used approaches would be extremely valuable for the discussion & descision on the best possible approach :+1:

**Comment by nponeccop** [comment]
Here is a procedure for a very minimal setup:

- `serverless create --template=aws-nodejs` - smallest of the templates
- `sed -i '/^#/d;/^$/d' serverless.yml` - delete comments
- `serverless deploy` - deploy hello: min-dev-hello and nothing else
- `jq '.Resources|map_values(.Type)' .serverless/cloudformation-template-*`
```json
{
  "ServerlessDeploymentBucket": "AWS::S3::Bucket",
  "ServerlessDeploymentBucketPolicy": "AWS::S3::BucketPolicy"
}
```

```json
{
  "ServerlessDeploymentBucket": "AWS::S3::Bucket",
  "ServerlessDeploymentBucketPolicy": "AWS::S3::BucketPolicy",
  "HelloLogGroup": "AWS::Logs::LogGroup",
  "IamRoleLambdaExecution": "AWS::IAM::Role",
  "HelloLambdaFunction": "AWS::Lambda::Function",
  "HelloLambdaVersionh6hfIK7ed0UWMENGZpRJoqXm2HhgzuIOzJaq1cD2nk": "AWS::Lambda::Version"
}
```

- No APIGatewayV2 is in sight, so go to APIGateway in AWS Console
- Create, Create Integrations, Lambda, {your region}, min-dev-hello, 2.0
- APIName = Serverless Minimal OpenAPI, as it's purely decorative
- Next
- Method=POST, leave everything else as is
- Next, Create
- You get "Successfully created API Serverless Minimal OpenAPI" green banner
- Go to Develop > Export in the sidebar
- rename the downloaded file to `openapi.yml` and put it next to `serverless.yml`
- Note that:
  - `info.title` is "Serverless Minimal OpenAPI". It isn't used as an identifier anywhere,
    but as a human-readable title in the ApiGateway console
  - `info.version` is autogenerated and doesn't seem to be used anywhere. You can use it
    as your API version in your public API documentation, or somethng like that
  - `servers` contain your `execute-api` domain name
  - stage name is `$default`, which lets you avoid that nasty `dev` stage name if you have
    only one stage. Deploying multiple stages in the same AWS account is almost useless,
    as e.g. DynamoDB doesn't have stages. And any project has multiple accounts, it's
    normal to have many if security is of any concern, as accounts are really
    the recommended security boundaries for PCI DSS/HIPAA/whatever. So I guess it's
    better to use `$default` in Serverless templates as it's a saner default, but I digress
  - `/min-dev-hello` in paths is a URI, you can change it independently from
    the lambda name
  - the lambda name `hello` from `serverless.yml` is in `x-amazon-apigateway-integration.uri`
  - `x-amazon-apigateway-integration.method` and `paths./min-dev-hello.post` are duplicates for
    a reason yet unclear to me
  - Now you can change openapi.yml and go to Develop > Reimport to change.

 The next step is to plug it into `serverless.yml` so it creates an OpenAPI-based ApiGatewayV2 by importing `openapi.yml` during deployment. And of course I'll create a repo for that at some point. It isn't very useful now because openapi.yml is polluted with AWS account id and ApiGateway Id.

**Comment by nponeccop** [comment]
Ok, here is an update. I managed to plug it. But the `openapi.yml` is still polluted at this point. Here is `serverless.yml`:

```yml
frameworkVersion: '2'
provider:
  name: aws
  runtime: nodejs12.x
  lambdaHashingVersion: 20201221
functions:
  hello:
    handler: handler.hello
resources:
  Resources:
    HttpApi:
      Type: AWS::ApiGatewayV2::Api
      Properties:
        Body: ${file(openapi.yml)}
    HttpApiStage:
      Type: AWS::ApiGatewayV2::Stage
      Properties:
        ApiId: !Ref HttpApi
        StageName: $default
        AutoDeploy: true
    HelloLambdaPermissionHttpApi:
      Type: AWS::Lambda::Permission
      Properties:
        FunctionName: !GetAtt HelloLambdaFunction.Arn
        Action: lambda:InvokeFunction
        Principal: apigateway.amazonaws.com
        SourceArn:
          Fn::Join:
            - ''
            - - 'arn:'
              - !Ref AWS::Partition
              - ':execute-api:'
              - !Ref AWS::Region
              - ':'
              - !Ref AWS::AccountId
              - ':'
              - !Ref HttpApi
              - '/*'
  Outputs:
    HttpApiId:
      Value: !Ref HttpApi
```

As you can see, I created by hand some of the resources Serverless normally creates:
- HttpApi - note that it's "empty" now, as everything is provided by the `openapi.yml`
- HttpApiStage
- HelloLambdaPermissionHttpApi

So far a course of action for Serverless can be:
- add a global `apiGateway.httpApiBody`. Note that ApiGatewayV1 also has something similar.
- the `httpApiBody` translates into the `Body` property of `AWS::ApiGatewayV2::Api`
- when the `Body` is specified, the behavior changes:
  - empty `httpApi` is allowed; it means the corresponding `LambdaPermissionHttpApi` is created and the rest is taken from `openapi.yml`
  - the `APIGatewayV2` is created without its default properties, as certain properties are prohibited when `Body` is present. The exact list of prohibited properties is yet to be determined.
  - `HttpApiStage` with its normal name taken from `stage`.

**Comment by nponeccop** [comment]
The above solution has a few flaws - there are parallel structures in `functions` in `serverless.yml` and in `openapi.yml`.

One possible solution is to add

```yaml
x-serverless.com:
   handler: handler.hello
```
to `openapi.yml`:

```yaml
frameworkVersion: '2'
provider:
  name: aws
  runtime: nodejs12.x
  lambdaHashingVersion: 20201221
  apiGateway:
     httpApiBody: ${file(openapi.yml)}
```

And in `openapi.yml` then we can have:

```yaml
openapi: "3.0.1"
info:
  title: "Serverless Minimal OpenAPI"
paths:
  /min-dev-hello:
    post:
      x-serverless.com:
        function: hello
        handler: handler.hello
```

**Comment by nponeccop** [comment]
I wrote a rather dirty `yaml-gen.js` script which converts the above `minimal.yml` into `openapi.yml`

```bash
npx -p yamljs json2yaml --depth 55 <(node yaml-gen.js <(npx -p yamljs yaml2json minimal.yml) us-east-1 1234567 min-dev) >openapi.yml
```

Unfortunately it needs 3 parameters now, to build the `uri` field: `us-east-1 1234567 min-dev`, region, account and lambda name prefix correspondingly. But it would allow me to test the proposed solution in a larger app prototype I'm developing. 

```js
const fs = require('fs')

const [,, fileName, region, account, lambdaPrefix] = process.argv

const minimal = JSON.parse(fs.readFileSync(fileName, 'utf8'))

const eachKvo = (o, f) => {
  for (let a of Object.keys(o)) f(a, o[a], o)
}

eachKvo(minimal.paths, (path, v, o) => {
  eachKvo(v, (method, v, o) => {
    const { handler, "function" : ff } = v['x-serverless.com']
    delete v['x-serverless.com']
    v['x-amazon-apigateway-integration'] = {
          "payloadFormatVersion": "2.0",
          "type": "aws_proxy",
          "httpMethod": method.toUpperCase(),
      "uri": `arn:aws:apigateway:${region}:lambda:path/2015-03-31/functions/arn:aws:lambda:${region}:${account}:function:${lambdaPrefix}-${ff}/invocations`,
          "connectionType": "INTERNET"
    }
  })
})

console.log(JSON.stringify(minimal))
```

**Comment by nponeccop** [comment]
Ok now a slightly better solution using variables coming from modules:

serverless.yml:
```yml
service: min
frameworkVersion: '2'
variablesResolutionMode: 20210326
unresolvedVariablesNotificationMode: error
configValidationMode: error
provider:
  name: aws
  runtime: nodejs12.x
  lambdaHashingVersion: 20201221
functions: ${file(minimal-openapi.js):functions}
resources:
  Resources:
    HttpApi:
      Type: AWS::ApiGatewayV2::Api
      Properties:
        Body: ${file(minimal-openapi.js):openapiJSON}
    HttpApiStage:
      Type: AWS::ApiGatewayV2::Stage
      Properties:
        ApiId: !Ref HttpApi
        StageName: $default
        AutoDeploy: true
    HelloLambdaPermissionHttpApi:
      Type: AWS::Lambda::Permission
      Properties:
        FunctionName: !GetAtt HelloLambdaFunction.Arn
        Action: lambda:InvokeFunction
        Principal: apigateway.amazonaws.com
        SourceArn:
          Fn::Join:
            - ''
            - - 'arn:'
              - !Ref AWS::Partition
              - ':execute-api:'
              - !Ref AWS::Region
              - ':'
              - !Ref AWS::AccountId
              - ':'
              - !Ref HttpApi
              - '/*'
  Outputs:
    HttpApiId:
      Value: !Ref HttpApi
```

minimal-openapi.js:
```js
const fs = require('fs')

const YAML = require('yamljs')
const assert = require('assert')

const fileName = 'minimal.yml'
const account = 208111775094

const file = fs.readFileSync(fileName, 'utf8')

const eachKvo = (o, f) => {
  for (let a of Object.keys(o)) f(a, o[a], o)
}

exports.openapiJSON = async ({ options, resolveConfigurationProperty }) => {
  const region = options.region
    || await resolveConfigurationProperty(['provider', 'region'])
    || 'us-east-1'
  const minimal = YAML.parse(file)
  const service = await resolveConfigurationProperty(['service'])
  const stage = await resolveConfigurationProperty(['provider', 'stage']) || 'dev'
  const lambdaPrefix = `${service}-${stage}`

  eachKvo(minimal.paths, (path, v, o) => {
    eachKvo(v, (method, v, o) => {
      const { "function" : ff } = v['x-serverless.com']
      delete v['x-serverless.com']
      v['x-amazon-apigateway-integration'] = {
            "payloadFormatVersion": "2.0",
            "type": "aws_proxy",
            "httpMethod": method.toUpperCase(),
        "uri": `arn:aws:apigateway:${region}:lambda:path/2015-03-31/functions/arn:aws:lambda:${region}:${account}:function:${lambdaPrefix}-${ff}/invocations`,
            "connectionType": "INTERNET"
      }
    })
  })
  return minimal
}

exports.functions = async ({ options, resolveConfigurationProperty }) => {
  const minimal = YAML.parse(file)
  const functions = {}

  eachKvo(minimal.paths, (path, v, o) => {
    eachKvo(v, (method, v, o) => {

      const vv = v['x-serverless.com']
      const ff = vv['function']
      delete vv['function']

      functions[ff] = vv
    })
  })
  return functions
}
```

**Comment by nponeccop** [comment]
An important finding is that AWS HTTP API ignores `requestBody` elements of OpenAPI 3 entirely. So there is no body validation  as per the OpenAPI spec on the AWS side, sorry. At least with HTTP APIs, I guess REST APIs and Websockets have a different set of features.

So far it seems the OpenAPI support in HTTP API is limited to:
- construction of `AWS::ApiGatewayV2::Integration` resources from the `x-amazon-apigateway-integration` extension
- construction of `AWS::ApiGatewayV2::Route` resources from the standard OpenAPI `paths` 

This elimination may or may not improve resource count, so the 200 resources per stack limitation is hit with more lambdas.

However, we still need the proposed `x-serverless.com` OpenAPI extension to eliminate the duplication between OpenAPI and Serverless completely. However, this extension is useful in other scenarios beyond the limited AWS-specific solution here.

If serverless parses OpenAPI itself, it can avoid `functions` even with other providers, which don't have built-in OpenAPI support! Even in the case of AWS, we can avoid `x-amazon-apigateway-integration` and generate all the necessary resources ourselves. We can switch it on and off with something like a `httpApi.nativeOpenApi: false` or `httpApi.generateBody: false`  setting.

**Comment by pgrzesik** [comment]
Hello @nponeccop - thanks a lot for all provided information. :bow: I set aside some time later this week in order to fully review your examples and rethink the whole potential OpenAPI support for HTTP API. I'll share my findings in this thread.

**Comment by nponeccop** [comment]
Meanwhile,  I created https://github.com/nponeccop/serverless-openapi-templates repo, with my current research and with the latest versions of the approaches above. I'm ready to transfer ownership if at some point you would like to have that in the `serverless` namespace/organization.

I think there are 3 possible ways of integration:
- generated parts of `serverless.yml` like I did
- a part of core `serverless`, as the `body` property is an integral feature of `httpApi`
- as a plugin

Having a real serverless template in addition to the integration, official or not, is a good idea anyway as OpenAPI is a game changer for many, and especially in the area of APIs. Typescript has some crazy libs for dynamic validation, so we'd likely need more than one template for JS and TS. Of course the research on templates is different from work on the integration, as it is on the language side.

**Comment by pgrzesik** [comment]
Hello @nponeccop - really sorry about the delay here. Unfortunatally currently we're focusing on other priorities, but if someone from community is interested in exploring potential ways to support OpenAPI definitions integration - we're open for help :100:

**Comment by nponeccop** [comment]
I didn't progress past the prototypes in that repo. We can work on improving the template.

This was for a real project, but I discovered that OpenAPI is relatively poorly supported throughout the infrastructure, so I decided to do it by hand.

As for the poor support, I mean for example request validation inside of AWS lambda. There are no good solutions I'm aware of. We both can work on things like this. I mean, well, there is middy middleware, but it's clunky to setup, and it doesn't even supports the precompiled validators of AJV, which are made for Lambda as they are very low footprint and cold startup time.

Now it's 2023 and we can use ChatGPT to speed up the prototyping :)
